### PR TITLE
Support additional language POC

### PR DESCRIPTION
**Pull request by eladiw**
This PR introduces a new language (this case hebrew) with two additional predefined recognizers.
IL ID and IL domain name.

What was changed:
1) analyzer already supported languages, however, now a recognizer by default supports 'all' languages and recognizers can override this.
2) when choosing a specific language for analyze, you get those language recognizers and also the recognizers which are not language specific.
3) spacy works only with english, still need to do, load other models (if needed) to support more languages. also since we've used the NLP engine for tokenization any language other than english doesnt support context support....

### CODE REVIEW

File: presidio-analyzer/presidio_analyzer/entity_recognizer.py [hunk 1]
```diff
@@ -128,13 +128,16 @@ def enhance_using_context(self, text, raw_results,
 
             word = text[result.start:result.end]
 
+            # Todo surronding words for unsupported
+            # languages doesnt work (due to tokenization)
```
Comment (omri374) [inline presidio-analyzer/presidio_analyzer/entity_recognizer.py]: Let's open a bug instead

File: presidio-analyzer/presidio_analyzer/analyzer_engine.py [hunk 2]
```diff
@@ -71,7 +71,10 @@ def Apply(self, request, context):
 
         entities = AnalyzerEngine.__convert_fields_to_entities(
             request.analyzeTemplate.fields)
+
+        logger.info("Getting language from request")
         language = AnalyzerEngine.get_language_from_request(request)
+        logger.info("Got language: '%s'", language)
```
Comment (omri374) [inline presidio-analyzer/presidio_analyzer/analyzer_engine.py]: Better to use f"Got language {language}" or .format() as these don't fail when param is not a string

File: presidio-analyzer/presidio_analyzer/pattern_recognizer.py [hunk 3]
```diff
@@ -16,7 +16,7 @@
 class PatternRecognizer(LocalRecognizer):
 
     def __init__(self, supported_entity, name=None,
-                 supported_language='en', patterns=None,
+                 supported_language='all', patterns=None,
```
Comment (omri374) [inline presidio-analyzer/presidio_analyzer/pattern_recognizer.py]: We once had a discussion about this and decided that each recognizer is language specific, even international ones like credit card. Main reason is the context which is always language specific. So even if we added French which has a spaCy model we can use, we would still need to figure out which context words to use.

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/il_domain_recognizer.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import tldextract
+
+from presidio_analyzer import Pattern, PatternRecognizer
+
+# pylint: disable=line-too-long
+IL_DOMAIN_REGEX = u'(www.)?[\u05D0-\u05EA]{2,63}.co.il'  # noqa: E501'  # noqa: E501
```
Comment (omri374) [inline presidio-analyzer/presidio_analyzer/predefined_recognizers/il_domain_recognizer.py]: What about org.il, gov.il etc.?

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/il_id_recognizer.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+# Todo (Task: TBD): fix context for unsupported languages
+CONTEXT = [
+    "תעודה",
+    "זהות",
+    "תעודת",
+    "ת.ז"
+]
+
+
```
Comment (omri374) [inline presidio-analyzer/presidio_analyzer/predefined_recognizers/il_id_recognizer.py]: Maybe for a consecutive PR but there's validation logic for Israeli IDs

File: presidio-analyzer/presidio_analyzer/pattern_recognizer.py (hunk 3 above)
Comment (eladiw) [reply presidio-analyzer/presidio_analyzer/pattern_recognizer.py]: but if we go this way we still get a strong set of multilanguage recognizers.
we can open this for discussion again or I can revert. both ok

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/il_domain_recognizer.py (hunk 4 above)
Comment (eladiw) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/il_domain_recognizer.py]: just an example for the POC... you are right

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/il_id_recognizer.py (hunk 5 above)
Comment (eladiw) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/il_id_recognizer.py]: will add.

File: presidio-analyzer/presidio_analyzer/entity_recognizer.py (hunk 1 above)
Comment (eladiw) [reply presidio-analyzer/presidio_analyzer/entity_recognizer.py]: opened

File: presidio-analyzer/presidio_analyzer/analyzer_engine.py (hunk 2 above)
Comment (eladiw) [reply presidio-analyzer/presidio_analyzer/analyzer_engine.py]: fixed

File: presidio-analyzer/presidio_analyzer/pattern_recognizer.py (hunk 3 above)
Comment (omri374) [reply presidio-analyzer/presidio_analyzer/pattern_recognizer.py:19]: I think we should revert. If someone wants to add real support for a new language, they'll be in trouble with all the context words definition.

Comment (eladiw) [reply presidio-analyzer/presidio_analyzer/pattern_recognizer.py:19]: ok.

Comment (eladiw) [reply presidio-analyzer/presidio_analyzer/pattern_recognizer.py:19]: Done @omri374

**Comment by omri374** [comment]
Code looks good. One suggestion is to add documentation on how to add a new language, with the considerations and limitations. Also mention that the IL ones are an example (they won't validate the ID for example and different types of domains would not get recognized)

**Comment by eladiw** [comment]
Done

File: docs/developing_recognizers.md [hunk 6]
```diff
@@ -45,46 +46,63 @@ See [this documentation](docs/custom_fields.md#via-code) on adding a new recogni
...
 #### Deep learning based methods
 
-Deep learning methods offer excellent detection rates for NER. 
-They are however more complex to train, deploy and tend to be slower than traditional approaches. 
+Deep learning methods offer excellent detection rates for NER.
+They are however more complex to train, deploy and tend to be slower than traditional approaches.
 When contributing a DL based method, the best option would be to create a sidecar container which is isolated from the presidio-analyzer container. On the `presidio-analyzer` side, one would extend the [RemoteRecognizer](presidio-analyzer/presidio_analyzer/remote_recognizer.py) class and implement the network interface between `presidio-analyzer` and the endpoint of the model's container.
 
+## Language support
+When developing a new recognizer, a supported language should be specified ('english' by default if not specified). When a request for analysis is made it includes the language as argument. Only recognizers supporting that language will be used (if any). A recognizer can support a single language only.
```
Comment (omri374) [inline docs/developing_recognizers.md:90]: either 'en' or English

File: docs/developing_recognizers.md [hunk 7]
```diff
@@ -45,46 +46,63 @@ See [this documentation](docs/custom_fields.md#via-code) on adding a new recogni
...
+see [Israeli web domains recognizer](../presidio-analyzer/presidio_analyzer/predefined_recognizers/il_domain_recognizer.py) as an example:
+
+```python
+def __init__(self):
+        patterns = [Pattern('IL Domain', IL_DOMAIN_REGEX, 0.5)]
+        super().__init__(supported_entity="IL_DOMAIN_NAME", patterns=patterns,
+                         context=CONTEXT, supported_language='he')
+```
+
+Important note: Presidio supports score improvment using surronding context words. However this mechanism is based on NLP tokenzier (SpaCy). In Presidio we are preloading only the English NLP model. This means that for languages other than English context support is not available.
```
Comment (omri374) [inline docs/developing_recognizers.md:104]: improvement
```suggestion
Important note: Presidio supports score improvement using surrounding context words. However this mechanism is based on NLP tokenzier (SpaCy). In Presidio we are preloading only the English NLP model. This means that for languages other than English context support is not available.
One can deploy Presidio with more than one spaCy model, but this is not officially supported. [Here is a list of supported languages by spaCy](https://spacy.io/usage/models)
```

File: docs/developing_recognizers.md [hunk 8]
```diff
@@ -45,46 +46,64 @@ See [this documentation](docs/custom_fields.md#via-code) on adding a new recogni
...
 #### Deep learning based methods
 
-Deep learning methods offer excellent detection rates for NER. 
-They are however more complex to train, deploy and tend to be slower than traditional approaches. 
+Deep learning methods offer excellent detection rates for NER.
+They are however more complex to train, deploy and tend to be slower than traditional approaches.
 When contributing a DL based method, the best option would be to create a sidecar container which is isolated from the presidio-analyzer container. On the `presidio-analyzer` side, one would extend the [RemoteRecognizer](presidio-analyzer/presidio_analyzer/remote_recognizer.py) class and implement the network interface between `presidio-analyzer` and the endpoint of the model's container.
 
+## Language support
+When developing a new recognizer, a supported language should be specified ('en' by default if not specified). When a request for analysis is made it includes the language as argument. Only recognizers supporting that language will be used (if any). A recognizer can support a single language only.
```
Comment (CodeRunRepeat) [inline docs/developing_recognizers.md:90]: Let's recommend to adopt the ISO 639-1 language code list.
https://www.iso.org/standard/22109.html
https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes

**Comment by VDCN12593** [comment]
Hi, anyone still interested in merging this? This makes a big difference for supporting other languages.

**Comment by omri374** [comment]
Hi @VDCN12593, yes definitely. There's another PR handling multiple languages which we are interested in merging first. There's currently a CI issue we are fixing before PRs can be approved.

File: presidio-analyzer/presidio_analyzer/nlp_engine/spacy_nlp_engine.py [hunk 9]
```diff
@@ -26,8 +26,16 @@ def process_text(self, text, language):
         """ Execute the SpaCy NLP pipeline on the given text
             and language
         """
-        doc = self.nlp[language](text)
-        return self.doc_to_nlp_artifact(doc, language)
+        if language in self.nlp:
+            doc = self.nlp[language](text)
+            return self.doc_to_nlp_artifact(doc, language)
+
+        # to do tokens and lemmas should be a simple tokenize so
```
Comment (balteravishay) [inline presidio-analyzer/presidio_analyzer/nlp_engine/spacy_nlp_engine.py:33]: this comment is not understood. is this a todo ?

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/il_id_recognizer.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
+from presidio_analyzer import Pattern
+from presidio_analyzer import PatternRecognizer
+# pylint: disable=line-too-long,abstract-method
+
+# Note: this recognizer is very simple and act mainly as an
```
Comment (balteravishay) [inline presidio-analyzer/presidio_analyzer/predefined_recognizers/il_id_recognizer.py:5]: if these are simple recognizers, is it worth committing them as code? can't they be created as "custom recognisers" and then added as documentation?

Comment (balteravishay) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/il_id_recognizer.py:5]: both this and ILDomainRecognizer

File: presidio-analyzer/tests/test_analyzer_engine.py [hunk 11]
```diff
@@ -154,15 +154,6 @@ def test_analyze_with_empty_text(self):
 
         assert len(results) == 0
 
-    def test_analyze_with_unsupported_language(self):
```
Comment (balteravishay) [inline presidio-analyzer/tests/test_analyzer_engine.py:157]: isn't there any more a concept of "unrecognised language"?

File: presidio-analyzer/tests/test_us_ssn_recognizer.py [hunk 12]
```diff
@@ -17,7 +17,7 @@ def test_valid_us_ssn_very_weak_match():
     assert_result_within_score_range(
         results[0], entities[0], 0, 10, 0, 0.3)
 
-    assert results[0].score != 0
+    assert results[1].score != 0
```
Comment (balteravishay) [inline presidio-analyzer/tests/test_us_ssn_recognizer.py:20]: why?

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/il_id_recognizer.py (hunk 10 above)
Comment (omri374) [reply presidio-analyzer/presidio_analyzer/predefined_recognizers/il_id_recognizer.py:5]: We mainly give this as an example, not as a feature. Putting them in JSON would likely be harder to figure out IMO.

**Comment by omri374** [comment]
Closing, this is supported in V2

**Pull request closed without merging**